sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/ResourceBoundedThreadExecutor.scala

```diff
@@ -0,0 +1,266 @@
+        futTask.task.releaseResourceCallback = () => mgr.releaseResource(futTask.task)
+        super.submit(futTask, null.asInstanceOf[T])
+      case _ =>
+        throw new UnsupportedOperationException("only accepts AsyncTask or RapidsFutureTask")
+    }
+  }
+
+  override def submit(r: Runnable): Future[_] = {
+    throw new UnsupportedOperationException("only accepts AsyncTask or RapidsFutureTask")
+  }
+
+  override protected def newTaskFor[T](fn: Callable[T]): RunnableFuture[T] = {
+    fn match {
+      case task: AsyncTask[_] =>
+        new RapidsFutureTask(task)
+      case f =>
+        throw new RuntimeException(s"Unexpected functor: ${f.getClass.getName}")
+    }
+  }
+
+  override protected def newTaskFor[T](r: Runnable, result: T): RunnableFuture[T] = {
+    r match {
+      case futTask: RapidsFutureTask[_] =>
+        futTask.asInstanceOf[RunnableFuture[T]]
+      case task: AsyncTask[_] =>
+        new RapidsFutureTask(task).asInstanceOf[RunnableFuture[T]]
+      case f =>
+        throw new RuntimeException(s"Unexpected runnable: ${f.getClass.getName}")
+    }
+  }
+
+  override def beforeExecute(t: Thread, r: Runnable): Unit = {
+    r match {
+      case fut: RapidsFutureTask[_] =>
+        mgr.acquireResource(fut.task, waitResourceTimeoutMs) match {
+          case s: AcquireSuccessful =>
+            fut.holdResource(s.elapsedTime)
+          case AcquireFailed =>
+            // bypass the execution via not holding the resource
+          case AcquireExcepted(exception) =>
+            logError(s"Invalid resource request for task ${fut.task}: ${exception.getMessage}")
+            // setException will unblock the corresponding waiting thread by failing it
+            fut.setException(exception)
+        }
+      case _ =>
+        throw new RuntimeException(s"Unexpected runnable: ${r.getClass.getName}")
+    }
+  }
+
+  override def afterExecute(r: Runnable, t: Throwable): Unit = {
+    r match {
+      case fut: RapidsFutureTask[_] =>
+        // Release the held resource if it was acquired.
+        if (fut.isHeldResource) {
+          // Release intermediately if the task supports that or if an exception occurred.
+          if (!fut.task.holdResourceAfterCompletion || t != null) {
+            fut.task.releaseResourceCallback()
+            fut.releaseResource()
+          } else {
+            fut.task.holdResource = true
+          }
+        }
+        // If the task failed to acquire enough resource, we bypass the execution and re-add it to
+        // the task queue with a priority penalty to avoid starvation.
+        if (t == null && !fut.isCompleted) {
+          // IMPORTANT: Ensure the priority penalty >=1000 in case the timeout task being
+          // polled recursively.
+          val priorPenalty = -retryPriorAdjust min -1e3f
+          fut.adjustPriority(priorPenalty, waitResourceTimeoutMs * 1000000L)
+          val holdingResource = fut.task.asInstanceOf[GroupedAsyncTask[_]].isHoldingResource
```

Comment:
fut.task.asInstanceOf[GroupedAsyncTask[_]].isHoldingResource -> how are we assured it's a GroupedAsyncTask here?

Reply:
If the task is a GroupedAsyncTask, then someone from its group must have successfully requested enough resource for all group members, right? How come the current task cannot "acquired the resource" ?

Reply:
> fut.task.asInstanceOf[GroupedAsyncTask[_]].isHoldingResource -> how are we assured it's a GroupedAsyncTask here?

Yes, this line was just for the previous debug. And it should be removed.

Reply:
if the task that is part of the group `GroupedAsyncTask` has to figure out the total `groupedHostMemoryBytes` before submitting to the controller => why should the controller know at all about grouped tasks? Why couldn't the group task just be a regular task that is implemented in the parquet code as a group of requests?

Reply:
> if the task that is part of the group `GroupedAsyncTask` has to figure out the total `groupedHostMemoryBytes` before submitting to the controller => why should the controller know at all about grouped tasks? Why couldn't the group task just be a regular task that is implemented in the parquet code as a group of requests?

After talking with @sperlingxx I understand groups better, I think with some comments we can make it clear.